File: Python/CountSubmatricesWithAllOnes.py
class Solution:
    def numSubmat(self, mat: list[list[int]]) -> int:
        m, n = len(mat), len(mat[0])
        h = [0] * n
        cnt = 0
        for i in range(m):
            for j in range(n):
                h[j] = h[j] + 1 if mat[i][j] else 0
            for j in range(n):
                mn, k = h[j], j
                while k >= 0 and mn:
                    mn = min(mn,h[k])
                    cnt += mn
                    k -= 1
        return cnt
    
    
s=Solution()
mat = [[1,0,1],[1,1,0],[1,1,0]]
print('Number of submatrices with all ones : ',s.numSubmat(mat))


# Input: mat = [[1,0,1],[1,1,0],[1,1,0]]
# Output: 13

# Input: mat = [[0,1,1,0],[0,1,1,1],[1,1,1,0]]
# Output: 24

File: Java/CountSubmatricesWithAllOnes.java
public class CountSubmatricesWithAllOnes {
    public int numSubmat(int[][] mat) {
        int m = mat.length, n = mat[0].length;
        int[] h = new int[n];
        int cnt = 0;

        for(int i=0; i<m; i++) {
            for(int j=0; j<n; j++) {
                h[j] = mat[i][j] == 1 ? h[j] + 1 : 0;
            }

            for(int j=0; j<n; j++) {
                int mn = h[j];
                for(int k=j; k>=0 && mn>0; k--) {
                    mn = Math.min(mn, h[k]);
                    cnt += mn;
                }
            }
        }
        return cnt;
    }

    public static void main(String[] args) {
        CountSubmatricesWithAllOnes obj = new CountSubmatricesWithAllOnes();

        int[][] mat = {
            {1, 0, 1},
            {1, 1, 0},
            {1, 1, 0}
        };

        int result = obj.numSubmat(mat);
        System.out.println("Number of submatrices with all ones: " + result);
    }
}


// Input: mat = [[1,0,1],[1,1,0],[1,1,0]]
// Output: 13

// Input: mat = [[0,1,1,0],[0,1,1,1],[1,1,1,0]]
// Output: 24